bpo-35017: Lib/socketserver, do not accept any request after shutdown

beledouxdenis opened the pull request:
Prior to this revision,
After the shutdown of a `BaseServer`,
the server accepted a last single request
if it was sent between the server socket polling
and the polling timeout.

This can be problematic for instance
for a server restart
for which you do not want to interrupt the service,
by not closing the listening socket during the restart.
One request failed because of this behavior.

Note that only one request failed,
following requests were not accepted, as expected.


<!-- issue-number: [bpo-35017](https://bugs.python.org/issue35017) -->
https://bugs.python.org/issue35017
<!-- /issue-number -->

We solved this issue in [odoo/odoo](https://github.com/odoo/odoo) by overriding _handle_request_noblock, in the below revision:
https://github.com/odoo/odoo/commit/dcb56412d667231c5427b0cd76cfc1c6c4b1dd3c#diff-7c4c00ad10cee1a79d1ed0d8b96f0bfaR140

We would like to avoid having to do so, as this is not a method which is supposed to be overridden.

beledouxdenis committed:
bpo-35017: `BaseServer` should not accept any request after shutdown (GH-9952)

Prior to this revision,
After the shutdown of a `BaseServer`,
the server accepted a last single request
if it was sent between the server socket polling
and the polling timeout.

This can be problematic for instance
for a server restart
for which you do not want to interrupt the service,
by not closing the listening socket during the restart.
One request failed because of this behavior.

Note that only one request failed,
following requests were not accepted, as expected.

tirkarthi commented on Misc/NEWS.d/next/Library/2018-10-19-09-32-07.bpo-35017.vAanVd.rst:
```diff
@@ -0,0 +1,2 @@
+After shutdown, :class:`BaseServer` accepted a last request if it was sent
```
As a personal suggestion I think this can be reworded to represent the behavior the PR proposes instead of the current behavior like below. I would wait for reviewer feedback though. 

```markdown
:class:`BaseServer` rejects the request sent between polling and polling 
timeout after shutdown.
```

Other than that it looks good to me. Thanks for the PR :)

vstinner replied:
Please mention the module name. Something like:

```
:meth:`socketserver.BaseServer.serve_forever` now exits immediately
if it's :meth:`~socketserver.BaseServer.shutdown` method is called
while it is polling for new events.
```

beledouxdenis replied:
I removed the former news entry and added a new one with your description. I couldn't have described this better.

vstinner commented on Lib/socketserver.py:
```diff
@@ -230,7 +230,7 @@ def serve_forever(self, poll_interval=0.5):
 
                 while not self.__shutdown_request:
                     ready = selector.select(poll_interval)
```
Why not "if not self.__shutdown_request: break"? Is it useful to call self.service_actions() even if we asked to stop the server? If you add this break, please add a short comment to explain how it can happen and add a "bpo-35017: " prefix to your comment to add the context.

beledouxdenis replied:
You are right. I will make the change right away.

vstinner commented on Lib/test/test_socketserver.py:
```diff
@@ -492,6 +492,37 @@ def shutdown_request(self, request):
+        thread_shutdown.start()
+
+        sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+        sock.settimeout(0.1)
```
It seems like your test rely on time for synchronization. I feel that it will fail if the sytem is busy or the threads are ordered differently depending on the OS :-(

I would either prefer to remove the test or write a test with better synchronization between the 3 threads.

cc @pablogsal

beledouxdenis replied:
The unit test was mainly to demonstrate there was an issue and how it could be triggered.

I hardly see how I could make the unit test not time based.

For the part
```python
sock.settimeout(0.1)
...
try:
    sock.recv(1024)
except socket.timeout:
    rejected = True
```
I could indeed maybe find another heuristic to determine the request was not accepted, but then remains the ```poll_interval``` timeout of the method
```python
def serve_forever(self, poll_interval=0.5):
```

If I increase this, this will take longer for the server to shutdown, and therefore the test will take time for an unnecessary reason.

I have no problem the unit test not to be merged. As explained, this was mainly to demonstrate the issue. Maybe it can be used as a reference in this pull request in case someone would like to know why this two lines patch was added.

I am opened to your suggestions if you have an idea to make this not time-based. Otherwise let me know and I will add a commit to remove the unit test.

pablogsal replied:
Indeed, I was able to make this test fail several times in our slowest buildbots :(

There is even the possibility that `shutdown` is called before `server_forver` starts the loop.

Also, shouldn't this be decorated with `@reap_threads`?

vstinner replied:
Please just remove the test. The fix is trivial, but I don't want to have to debug race condition in this new functional test.

beledouxdenis replied:
I fully agree. I just removed the unit test case.

beledouxdenis committed:
bpo-35017: remove the unit test as it's time-based

This test can fail on slow buildbots,
because it relies on a time condition

The unit test can be safely removed,
the fix being quite trivial.

beledouxdenis committed:
bpo-35017: reword the news to represent the new behavior

The news message now represents what the revision
is changing regarding the behavior rather
than describing the current unexpected behavior.

vstinner commented on Lib/socketserver.py:
```diff
@@ -230,6 +230,10 @@ def serve_forever(self, poll_interval=0.5):
 
                 while not self.__shutdown_request:
                     ready = selector.select(poll_interval)
+                    # bpo-35017: request arriving after the server shutdown
```
I'm confused by this comment. I suggest something like:

"bpo-35017: shutdown() has been called from a different thread while select() was polling. Don't handle the request, exit immediatly."

... or maybe remove the comment, I don't know :-)

odony replied:
If I may suggest a shorter version with the same meaning:
```python
# bpo-35017: shutdown() called during select(), exit immediately
``` 
Though I think the code is quite self-explanatory, and I would omit the comment ;-)

beledouxdenis replied:
Let me know If I should use @odony  reword, or if you prefer removing the comment completely.

vstinner replied:
"# bpo-35017: shutdown() called during select(), exit immediately" is better than my long boring comment :)

tirkarthi replied:
@vstinner Off topic but just as a pro-tip now you can suggest changes in GitHub UI that can be accepted and committed from UI when the contributor clicks "Accept suggestion" that might help during rewording :) Ref : https://twitter.com/github/status/1054794897448939520 

Edit : Doesn't work with multiple lines it seems. Sorry for the noise.

```suggestion
                    # bpo-35017: shutdown() called during select(), exit immediately
```

beledouxdenis committed:
bpo-35017: reword inline comment to be shorter and cleaner
